### PR TITLE
Implement HTMLMediaElement defaultPlaybackRate and playbackRate

### LINKED ISSUE #22293: Implement HTMLMediaElement playbackRate and defaultPlaybackRate attributes
Spec: https://html.spec.whatwg.org/multipage/media.html#dom-media-playbackrate

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #22293 

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/22449)

### CODE REVIEW

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: Right now we are only storing and retrieving the playback rate value from this local member. We still need to pass this value to the actual player, using the API that you introduced in https://github.com/servo/media/pull/168. You need to update servo-media (`./mach cargo-update -p servo-media`) and make use of the new API here.

**Review by ferjm — CHANGES_REQUESTED** [review]

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: We could add a new `queue_ratechange_event` method, so we don't duplicate these lines.

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: Or even better, since this is a pattern that is repeated all over this file, we could have a `fn queue_media_event(&self, event: &str)` method. But it's ok to do this in a follow-up.

Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: I chose the first method so as to keep the changes as self-contained as possible. Once this PR is done I'll refactor the whole file using the second method.

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: Use `warn!` instead of `eprintln!` here, please. (I know that there are other uses of `eprintln!` in this file, but we should switch all of them to `warn!` or `error!` at some point)

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: Let's do this slightly different to avoid duplicated `ratechange` events:

```rust
if self.playbackRate.get() == r {
    self.queue_ratechange_event();
} else {
    self.playbackRate.set(r);
}
```

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: According to the [gst-player docs](https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-bad-libs/html/GstPlayer.html#GstPlayer--rate) the allowed range for the `rate` property is [-64,64] and the [spec](https://html.spec.whatwg.org/multipage/media.html#dom-media-playbackrate) expects us to throw a `NotSupportedError` exception if we try setting a value out of the allowed range.

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: We receive the `Player::RateChanged` event from the media backend once the new rate has been actually set and we are already queueing the `ratechange` event in the `Player::RateChanged` handler, so there's no need to queue it here (we would get duplicated events).

**Review by ferjm — CHANGES_REQUESTED** [review]
Almost there :)

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: Also according to the spec, we should only change the rate if we are [potentially playing](https://html.spec.whatwg.org/multipage/media.html#potentially-playing)

Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: I looked at how Gecko [does](https://github.com/mozilla/gecko-dev/blob/47edbd91c43db6229cf32d1fc4bae9b325b9e2d0/dom/html/HTMLMediaElement.cpp#L5772) this check and noticed that it misses some steps in the process. I am not really sure how to go about implementing it, could you, please, give me some guidelines?

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: Which tests are timing out? Could you update the PR with your latest changes, please?

If it is timing out it may be because we are not getting the `Player::RateChanged` in some cases or because there is something else wrong.

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: Yeah, we don't have all the required pieces to completely implement all the checks, but we can at least leave the corresponding placeholders:

- We depend  on #22348 to get the `Ended` method.

- I suspect we depend on the MediaSession API to be able to implement the `is_paused_for_user_interaction` method.

- We certainly depend on #22314 to implement `is_paused_for_in_band_content`. Let's just leave the placeholders for now.

I would do something like:

```rust
fn is_potentially_playing(&self) -> bool {
    !self.paused.get() &&
    // FIXME: We need https://github.com/servo/servo/pull/22348
    //              to know whether playback has ended or not
    // !self.Ended() &&
    self.error.get().is_none() &&
    !self.is_blocked_media_element()
}

fn is_blocked_media_element(&self) -> bool {
    self.readyState.get() <= ReadyState::HaveCurrentData ||
    self.is_paused_for_user_interaction() ||
    self.is_paused_for_in_band_content()
}

fn is_paused_for_user_interaction(&self) -> bool {
    // FIXME: we will likely be able to fill this placeholder once (if) we
    //             implement the MediaSession API.
    false
}

fn is_paused_for_in_band_content(&self) -> bool {
     // FIXME: we will likely be able to fill this placeholder once (if) we implement
     //             https://github.com/servo/servo/issues/22314
    false
}
```

File: components/script/dom/htmlmediaelement.rs
Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: Here are the tests that time out:
image

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: It seems that the problem is in servo-media. Setting the playback rate [here](https://github.com/servo/media/blob/master/backends/gstreamer/src/player.rs#L392) completely breaks a/v playback. It even [failed on Travis](https://travis-ci.org/servo/media/jobs/467472875). Sorry, I should have catched that on the review.

I still don't know why that fails. I will investigate more tomorrow.

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: I found the reason why this was not working. We were creating a deadlock while trying to send the RateChanged event. The good news is that we don't need this event at all. I mistakenly thought that the rate change signal was triggered when the actual rate change was effective and that the player could change its playback rate for other reasons other than the client asking for that change, but apparently none of my assumptions were correct. The signal is triggered only when the rate **property** is changed and the playback rate will only change if the client instructs the player to change it and so we don't need to notify the client about that change.

I am removing the signal handler and the Player::RateChanged event in https://github.com/servo/media/pull/175.

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: @georgeroman https://github.com/servo/media/pull/175 has already landed, so we can update this PR to remove the Player::RateChanged logic. Thanks for your pacience.

Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: Great! I updated the code accordingly and now the tests pass.

File: components/script/dom/htmlmediaelement.rs
Comment (ferjm) [inline components/script/dom/htmlmediaelement.rs]: One last change! Let's set the property and queue the `ratechange` event only if setting the player rate succeeds.

**Review by ferjm — APPROVED** [review]
Looks great! r=me with the last comment addressed. Thanks!

(Our CI env is on fire now, so we need to wait for #22493 to land before being able to merge this one)

**Comment by ferjm** [comment]
Oh, you will likely need to update the test expectations for https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/interfaces.html as well.

@bors-servo try=wpt

File: components/script/dom/htmlmediaelement.rs
Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: With this change in place, it looks like the tests timeouts are back.

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: Ok. I think this may be a misinterpretation (or actually a lack of clarity) of the spec.

So the spec says [here](https://html.spec.whatwg.org/multipage/media.html#dom-media-playbackrate):

> Set playbackRate to the new value, and if the element is potentially playing, change the playback speed.

Changing the playback speed (`self.player.set_rate()`) depends on the element being potentially playing. 

But then it says as well:

> When the defaultPlaybackRate or playbackRate attributes change value (either by being set by script or by being changed directly by the user agent, e.g. in response to user control) the user agent must queue a task to fire an event named ratechange at the media element.

It does not say anything about firing the `ratechange` event only if the element is potentially playing. So we will need to change the attribute value (`self.playbackRate.set(*value)`) and fire the event (`self.queue_ratechange_event()`) **even if the element is not potentially playing**.

The problem with allowing the `playbackRate` attribute change when the media element is not playing is that we will be missing the actual rate change when the media starts playing. So to fix that issue we will need to apply the current `playbackRate` attribute value to the player (`self.player.set_rate(self.playbackRage.get())`) right before the media starts playing.

Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: Thank you for the explanations! I opened an issue here: [https://github.com/whatwg/html/issues/4256](https://github.com/whatwg/html/issues/4256)

**Review by ferjm — APPROVED** [review]
Great work! Thanks :)

**Comment by ferjm** [comment]
@bors-servo retry

- #22007

**Comment by ferjm** [comment]
@bors-servo retry
- #22007

**Comment by ferjm** [comment]
Hmm ... `/html/semantics/embedded-content/media-elements/event_timeupdate.html` intermittently passes because gstreamer performs a couple of initial seek requests with ogg files (this does not happen with other containers) and we are triggering the `timeupdate` event when a seek is completed.
The problem is that ogg playback is not very reliable on gstreamer, and I am forcing the usage of mp4 for the WPT suite in #22477.
Since the test is mostly testing the _time marches on_ algorithm rather than `timeupdate` events fired from other places and we are not implementing this algorithm yet, I think we should disable this test and `/html/semantics/embedded-content/media-elements/event_timeupdate_noautoplay.html` for now until I figure out how to merge #22477.

File: components/script/dom/htmlmediaelement.rs
Comment (georgeroman) [reply components/script/dom/htmlmediaelement.rs]: It seems that updating the playback rate of the media element was implicitly mentioned in another paragraph of the API: https://github.com/whatwg/html/issues/4256#issuecomment-453116074

Comment (ferjm) [reply components/script/dom/htmlmediaelement.rs]: Ah, I see! Thanks for following up on that. The code fully reflects what the spec says then. Thanks!

**Pull request merged**
